File: javascript/protocol/socket.js
Faye.Server.Socket = Faye.Class({
  initialize: function(server, socket, request) {
    this._server  = server;
    this._socket  = socket;
    this._request = request;
  },

  send: function(message) {
    this._server.pipeThroughExtensions('outgoing', message, this._request, function(pipedMessage) {
      if (this._socket)
        this._socket.send(Faye.toJSON([pipedMessage]));
    }, this);
  },

  close: function() {
    if (this._socket) this._socket.close();
    delete this._socket;
  }
});

File: lib/faye/protocol/socket.rb
module Faye
  class Server

    class Socket
      def initialize(server, socket, env)
        @server = server
        @socket = socket
        @env    = env
      end

      def send(message)
        @server.pipe_through_extensions(:outgoing, message, @env) do |piped_message|
          @socket.send(Faye.to_json([piped_message])) if @socket
        end
      end

      def close
        @socket.close if @socket
        @socket = nil
      end
    end

  end
end
